distribution/xfer/download.go

```diff
@@ -81,6 +97,104 @@ type DownloadDescriptorWithRegistered interface {
 	Registered(diffID layer.DiffID)
 }
 
+// CollectGarbage removes any cached layers associated with downloads that are
+// still in progress
+func (ldm *LayerDownloadManager) CollectGarbage() error {
+	// we just flush the whole resumption cache and then save it
+	ldm.cachedLayerRefsLock.Lock()
+	for _, l := range ldm.cachedLayerRefs {
+		layer.ReleaseAndLog(ldm.layerStore, l)
+	}
+	ldm.cachedLayerRefs = make(map[string]layer.Layer)
+	ldm.cachedLayerRefsLock.Unlock()
+	ldm.saveLayerRefs()
+
+	return ldm.tm.CollectGarbage()
+}
+
+func (ldm *LayerDownloadManager) loadLayerRefs() {
+	data, err := ioutil.ReadFile(ldm.imagePullCheckpointsPath)
+	if err != nil {
+		logrus.Warnf("failed to read image pull checkpoint: %s", err)
+	}
+	checkpoint := map[string]string{}
+	err = json.Unmarshal(data, &checkpoint)
+	if err != nil {
+		logrus.Warnf("failed to unmarshal pull checkpoint: %s", err)
+	}
+	ldm.cachedLayerRefsLock.Lock()
+	defer ldm.cachedLayerRefsLock.Unlock()
+	for key, chainID := range checkpoint {
+		layer, err := ldm.layerStore.Get(layer.ChainID(chainID))
```

Comment:
Yeah, I thought about that. It would be simpler but it would re-extract layers redundantly. I didn't realize how complex this would get with the current design. I'll try making a separate PR with that approach.